src/librustdoc/docfs.rs

```diff
@@ -0,0 +1,124 @@
+//! Rustdoc's FileSystem abstraction module.
+//!
+//! On Windows this indirects IO into threads to work around performance issues
+//! with Defender (and other similar virus scanners that do blocking operations).
+//! On other platforms this is a thin shim to fs.
+//!
+//! Only calls needed to permit this workaround have been abstracted: thus
+//! fs::read is still done directly via the fs module; if in future rustdoc
+//! needs to read-after-write from a file, then it would be added to this
+//! abstraction.
+
+use errors;
+
+use std::cell::RefCell;
+use std::fs;
+use std::io;
+use std::path::Path;
+use std::sync::Arc;
+use std::sync::mpsc::{channel, Receiver, Sender};
+
+macro_rules! try_err {
+    ($e:expr, $file:expr) => {{
+        match $e {
+            Ok(e) => e,
+            Err(e) => return Err(E::new(e, $file)),
+        }
+    }};
+}
+
+pub trait PathError {
+    fn new<P: AsRef<Path>>(e: io::Error, path: P) -> Self;
+}
+
+pub struct ErrorStorage {
+    sender: Sender<Option<String>>,
+    receiver: Receiver<Option<String>>,
+    nb_threads: RefCell<usize>,
+}
+
+impl ErrorStorage {
+    pub fn new() -> ErrorStorage {
+        let (sender, receiver) = channel();
+        ErrorStorage {
+            sender,
+            receiver,
+            nb_threads: RefCell::new(0),
+        }
+    }
+
+    /// Prints all stored errors. Returns the number of printed errors.
+    pub fn write_errors(&self, diag: &errors::Handler) -> usize {
+        let mut printed = 0;
+        let nb_threads: usize = *self.nb_threads.borrow();
+
+        if nb_threads > 0 {
+            // We limit the wait to 30 seconds maximum.
+            for _ in 0..nb_threads {
```

Comment:
This should just use self.receiver.iter() and not bother counting nb_threads:

 - recv() will error (as oppose to hanging) if all the senders have been dropped
 - panic will unwind the worker thread and drop the closure, dropping the moved send clone

Switching to iter will behave the same, but you'll have less code as the all-senders-hungup cases will be handled transparently for you.

Reply:
```
for msg in self.receiver.iter() {
    if let Some(ref error) = msg {
        diag.struct_err(&error).emit();
        printed += 1;
    }
}
printed
````